### PR TITLE
accept tripple dash file URIs as valid links

### PR DESCRIPTION
**Pull request by tjex**
PR to fix zk-org/zk#250. Links such as `[title](file:///file.go)` were returning an
lsp error "not found". The LSP does not actually check for the existence of
files via this format. Instead it was a symptom of the markdown link regex
checking, whereby the link was being understood as an internal markdown
document. Zk / lsp then tried to find the note `file:///file.example`, which it
if course couldn't, and so it was throwing the "not found" error.

This sollution copies the logic already present for embedded images, which is to
pick a defining syntactic element of the link (in this case, the third
occurrence of a '/' rune), and to skip that link with `continue`, effectively
ignoring it.

### CODE REVIEW

**Comment by tjex** [comment]
sorry for the formatting.. looks like the github cli is sending as formatted raw text rather than markdown compatible text... bit of a nish nish.

File: internal/adapter/lsp/document.go
Comment (mcDevnagh) [inline internal/adapter/lsp/document.go:236]: this is a very over-eager match
this matches the **/** here: file:/**/**/, but it would also match the following: ./dir1/file

Also, I don't think this is a good way to use the result of `FindAllStringSubmatchIndex`
- For the `markdownLinkRegex` (`\[([^\]]+?[^\\])\]\((.+?[^\\])\)`), it returns a slice of int slices
    - Each slice within the slice corresponds to a match
    - Each match will contain exactly 6 integers:
        - the starting index of the entire match (inclusive)
        - the ending index of the entire match (exclusive)
        - the starting index of the *bracket* capture group (inclusive)
        - the ending index of the *bracket* capture group (exclusive)
        - the starting index of the *parenthesis* capture group (inclusive)
        - the ending index of the *parenthesis* capture group (exclusive)
    - `match[3]` then corresponds to the index ending the _bracket_ capture group, which seems odd, since we're interested in the *parenthesis* capture group
        - `match[4]` is the one we should use 

Thus, I think we should match against the entire [scheme](https://en.wikipedia.org/wiki/File_URI_scheme)--`file://`--starting at `match[4]`. This is probably best done in a helper function, first checking that `match[5] - match[4]` is large enough for `file://` (8, since we need at least one character after the scheme to be a valid URI), then iterating from `match[4]` to `match[5]` matching each character along the way. If any character doesn't match, we return a `false` early.

Also, I know we're not explicitly supporting windows, but this will cause an obvious portability issue:
- POSIX has the third slash in `file:///` correspond to an actual file: the root file
- Windows seems to append an additional `/`, since `file:///c:` corresponds to the C drive, and `file:///` corresponds to nothing

I don't think we necessary need to support `file://` on Windows with this PR, but I am curious as to the result of this on Windows

File: internal/adapter/lsp/document.go
Comment (mcDevnagh) [inline internal/adapter/lsp/document.go:15]: why the reordering? Did gofmt do this?

**Review by mcDevnagh — CHANGES_REQUESTED** [review]

File: internal/adapter/lsp/document.go
Comment (tjex) [reply internal/adapter/lsp/document.go:15]: yes. Same as the spaces / tabs you pointed out above... I went back to double check and ran `go fmt` again.

File: internal/adapter/lsp/document.go
Comment (tjex) [reply internal/adapter/lsp/document.go:236]: Ahhh. I thought the ints followed the sequence in which they were found... as in 0 -> n from `[` -> `)`.
I'll have a proper look through all this in the coming days hopefully. But I'm on holiday for a month from Thursday!

Thanks so much for all the detail and thoroughness! 🙏

Comment (mcDevnagh) [reply internal/adapter/lsp/document.go:236]: np. it took a lot of testing for me to figure out what exactly `FindAllStringSubmatchIndex` does. It's documentation leaves a lot to be desired

File: internal/adapter/lsp/document.go
Comment (mcDevnagh) [reply internal/adapter/lsp/document.go:15]: sounds good. I'm fine with unrelated changes like this if it wasn't done properly before (i.e. `go fmt` wasn't run), but if its just a matter of preference, I'd rather keep the git blame clean. Looks like we're good to keep this

Comment (tjex) [reply internal/adapter/lsp/document.go:15]: Nice, me too. I appreciate that level of attention / scrutiny 🔥

File: internal/adapter/lsp/document.go
Comment (tjex) [reply internal/adapter/lsp/document.go:236]: @mcDevnagh 

Would this work, or is it in an ok direction?

```go
package main

import (
	"regexp"
)

const testCase = "[example](file:///foo.go)"

var markdownLinkRegex = regexp.MustCompile(`\[([^\]]+?[^\\])\]\((.+?[^\\])\)`)
var fileURIregex = regexp.MustCompile(`file:///`)

func main() {

	var result string
	for _, match := range markdownLinkRegex.FindAllStringSubmatchIndex(testCase, -1) {
		if match[5]-match[4] >= 8 {
			// get the file url portion, existing within the parentheses
			linkURL := testCase[match[4]:match[5]]
			// limit to first 8 characters as per 'file:///' scheme
			result = linkURL[:8]
		}
		// check if the first 8 characters of the URL match 'file:///'
		if fileURIregex.MatchString(result) {
			// if they match, we continue, meaning we DO NOT consider it an
			// internal note url, thus avoiding the lsp error that the note doesn't exist
			continue
		}
	}

}
```

Comment (tjex) [reply internal/adapter/lsp/document.go]: I added this approach into document.go and tested. waiting for review.

**Review by jurica — COMMENTED** [review]
Looks good to me.

**Comment by tjex** [comment]
@mcDevnagh are you around to approve your reviewed changes (merging is blocked until you review)? Otherwise I'll dismiss and ask someone else to approve (just feels rude to dismiss until I try to ping you here as well).

**Comment by mcDevnagh** [comment]
I'll take a look

**Comment by tjex** [comment]
> Looks good to me.

@jurica , I think mcDevangh is a bit too snowed under to re-review. I've dismissed his review so that you can pass the review yourself. I'd like to move on to tackling the other bugs, which will be working in the same file as this PR.

**Review by jurica — APPROVED** [review]

**Comment by jurica** [comment]
> > Looks good to me.
> 
> @jurica , I think mcDevangh is a bit too snowed under to re-review. I've dismissed his review so that you can pass the review as well, incase he needs more time. Just caus I'd like to move on to tackling the other bugs this long weekend, which will be working in the same file as this PR.

Sure if it's holding you back from doing more. I didn't want to merge it right away last week to give others the chance to also review...

**Comment by mcDevnagh** [comment]
sorry for dropping the ball on this one! Thank you @jurica for picking up my slack. For what it's worth, code looks good :)

**Pull request merged**
